### PR TITLE
chore: add webhook tests for notification subsystem

### LINKED ISSUE #5: Add detailed unit tests for SMTP & webhook notification dispatchers

### PR DESCRIPTION
**Pull request by dannykopping**
Closes https://github.com/coder/team-coconut/issues/5

### CODE REVIEW

File: coderd/notifications/metrics_test.go [hunk 1]
```diff
@@ -263,9 +263,10 @@ func TestPendingUpdatesMetric(t *testing.T) {
 	interceptor.proceed.Broadcast()
 
 	// Validate that the store synced the expected number of updates.
-	require.Eventually(t, func() bool {
+	require.Eventuallyf(t, func() bool {
 		return syncer.sent.Load() == 1 && syncer.failed.Load() == 1
-	}, testutil.WaitShort, testutil.IntervalFast)
+	}, testutil.WaitLong, testutil.IntervalFast,
+		"sent: %d, failed: %d", syncer.sent.Load(), syncer.failed.Load())
```
Comment (dannykopping) [inline coderd/notifications/metrics_test.go:269]: Drive-by: this test flaked once but I couldn't find the reason after a quick look, so I'm extending the wait time a bit and adding some debugging.

File: coderd/notifications/dispatch/webhook_test.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		expectSuccess   bool
+		expectRetryable bool
+		expectErr       string
+	}{
+		{
+			name: "successful",
+			serverFn: func(msgID uuid.UUID, w http.ResponseWriter, r *http.Request) {
+				var payload dispatch.WebhookPayload
+				err := json.NewDecoder(r.Body).Decode(&payload)
+				require.NoError(t, err)
```
Comment (mafredri) [inline coderd/notifications/dispatch/webhook_test.go:57]: Since this is running in a goroutine, let's change these to `assert`.

> FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test.

File: coderd/notifications/dispatch/webhook_test.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		},
+		{
+			name: "non-200 response",
+			serverFn: func(_ uuid.UUID, w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusInternalServerError)
+			},
+			expectSuccess:   false,
+			expectRetryable: true,
+			expectErr:       "non-2xx response (500)",
+		},
```
Comment (mafredri) [inline coderd/notifications/dispatch/webhook_test.go:91]: Perhaps we could add a test that sends bad data (wrong format) too, just to have the case covered?

**Review by mafredri — APPROVED** [review]
I'd like to see all goroutine `require`s changed to `assert`, but other than that, LGTM!

File: coderd/notifications/dispatch/webhook_test.go (hunk 2 above)
Comment (dannykopping) [reply coderd/notifications/dispatch/webhook_test.go:57]: Oh you're right! The `http.Server` serves in a goroutine, whoops.

File: coderd/notifications/dispatch/webhook_test.go (hunk 3 above)
Comment (dannykopping) [reply coderd/notifications/dispatch/webhook_test.go:92]: It's not actually possible, the webhook dispatch sends a well-formed request.

See `func (w *WebhookHandler) dispatch`

**Review by mtojek — COMMENTED** [review]
post-merge approval
